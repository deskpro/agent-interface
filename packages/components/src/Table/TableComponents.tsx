import * as React from "react";
import classNames from "classnames";

import "@deskpro/agent-interface-style/dist/components/dp-Tableview.css";

export const TableView: React.FC<React.ComponentProps<"div">> = ({
  className,
  ...props
}) => <div className={classNames("dp-Tableview", className)} {...props} />;

export type RowProps = React.ComponentProps<"div"> & {
  even?: boolean;
  expanded?: boolean;
  grouped?: boolean;
};
export const Row: React.FC<RowProps> = ({
  className,
  even = false,
  expanded = false,
  grouped = false,
  ...props
}) => (
  <div
    className={classNames("dp-TableRow", className, {
      "is-even": even,
      "is-selected": expanded,
      "dp-Filter": grouped
    })}
    {...props}
  />
);

export const HeaderRow: React.FC<RowProps> = ({ className, ...props }) => (
  <Row className={classNames("dp-TableControl", className)} {...props} />
);

export const Cell: React.FC<React.ComponentProps<"div">> = props => (
  <div {...props} />
);

export type HeaderProps = React.ComponentProps<"div"> & {
  canSortBy?: boolean;
  sorted?: boolean;
  sortedIndex?: number;
  sortedDesc?: boolean;
  sortByProps?: any;
};

export const Header: React.FC<HeaderProps> = ({
  canSortBy = false,
  sorted = false,
  sortedDesc,
  sortedIndex,
  sortByProps = {},
  children,
  className,
  ...props
}) => (
  <Cell {...props} className={classNames("dp-TableItem", className)}>
    {children}
    {canSortBy && (
      <span
        {...sortByProps}
        className={classNames("dp-Icon dp-arrowGroup", sortByProps.className, {
          "is-asc": sorted && !sortedDesc,
          "is-desc": sorted && sortedDesc
        })}
      />
    )}
  </Cell>
);
