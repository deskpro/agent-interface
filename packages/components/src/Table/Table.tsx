import * as React from "react";
import {
  actions,
  useTable,
  useColumns,
  useRows,
  // useGroupBy,
  useSortBy,
  useExpanded,
  useFlexLayout,
  useTableState,
  HeaderColumn,
  EnhancedColumn as ReactTableEnhancedColumn
} from "react-table";
import classNames from "classnames";

import { TableView, Header, Cell, Row, HeaderRow } from "./TableComponents";
import useItemSelection from "../ListPane/useItemSelection";
import { ListItem, ListItemGroup } from "../ListPane/types";
import Checkbox from "../inputs/Checkbox/Checkbox";
import Arrow from "../Button/Arrow";

interface EnhancedColumn extends ReactTableEnhancedColumn {
  headerClassName?: string;
}

export type TableProps = {
  data: ListItem[];
  columns: any[];
  defaults?: {
    pageSize?: number;
  };
  hoverable?: boolean;
  expandable?: boolean;
  renderExpandedRow?: (row: ListItem) => React.ReactNode;
  checkable?: boolean;
  onCheckChange?: (selection: React.Key[]) => void;
  onSort?: (sortBy?: { id: string; desc: boolean }) => void;
};

const defaultGroup: ListItemGroup = { id: "DEFAULT", title: "default" };

const Table: React.FC<TableProps> = ({
  hoverable = true,
  expandable = false,
  checkable = false,
  onCheckChange,
  renderExpandedRow,
  defaults = {},
  columns,
  onSort,
  data,
  ...props
}) => {
  const expandColumn = {
    Header: "",
    id: "expander",
    width: 20,
    Cell: ""
  };
  const checkboxColumn = {
    Header: "",
    id: "checkbox",
    width: 40,
    Cell: ""
  };
  const instance = useTable(
    {
      ...props,
      data,
      state: useTableState(defaults, undefined, {
        reducer: (_, newState, type) => {
          if (type === actions.sortByChange && typeof onSort === "function") {
            const [sortBy] = newState.sortBy;
            onSort(sortBy);
          }
          return newState;
        }
      }),
      columns: [expandable && expandColumn, checkable && checkboxColumn]
        .concat(columns)
        .filter(Boolean) as HeaderColumn[],
      manualSorting: typeof onSort === "function",
      disableMultiSort: true
    },
    useColumns,
    useRows,
    useSortBy,
    useExpanded,
    useFlexLayout
  );

  const {
    getTableProps,
    headerGroups: headerRows,
    rows,
    prepareRow
  } = instance;

  const {
    selection,
    toggleItemSelection,
    toggleGroupSelection,
    getGroupCheckState,
    handleSelectAll,
    handleSelectNone
    // handleSelectInverse
  } = useItemSelection(data);
  React.useEffect(() => {
    if (onCheckChange) {
      onCheckChange(selection);
    }
  }, [selection, onCheckChange]);

  const renderRow = (row, index, rowsArr) => {
    if (!row) {
      return null;
    }
    prepareRow(row);

    const group = row.original.group || defaultGroup;
    const prevGroup =
      index !== 0 ? rowsArr[index - 1].original.group || defaultGroup : null;
    const shouldRenderGroupHeader =
      group.id !== defaultGroup.id &&
      (prevGroup === null || prevGroup.id !== group.id);
    const groupCheckState = getGroupCheckState(group.id);

    const rowProps = row.getRowProps();

    const renderedRow = (
      <Row {...rowProps} expanded={row.isExpanded}>
        {row.cells.map(cell => (
          <Cell {...cell.getCellProps({ className: cell.column.className })}>
            {cell.column.id === "expander" && (
              <Arrow
                onClick={() => row.toggleExpanded()}
                isActive={row.isExpanded}
              />
            )}
            {cell.column.id === "checkbox" && (
              <Checkbox
                id={row.original.id}
                checked={selection.includes(row.original.id)}
                onChange={() => toggleItemSelection(row.original.id)}
              />
            )}
            {cell.render("Cell")}
          </Cell>
        ))}
      </Row>
    );

    if (row.isExpanded || shouldRenderGroupHeader) {
      return (
        <React.Fragment key={rowProps.key}>
          {shouldRenderGroupHeader && (
            <Row {...rowProps} key={`group_${group.id}`} grouped>
              {expandable && <Cell {...row.cells[0].getCellProps()} />}
              {checkable && (
                <Cell {...row.cells[expandable ? 1 : 0].getCellProps()}>
                  <Checkbox
                    id={group.id}
                    checked={!!groupCheckState}
                    undef={groupCheckState === "undef"}
                    onChange={() => toggleGroupSelection(group.id)}
                  />
                </Cell>
              )}
              <Cell className="dp-Card-title" style={{ flex: "0 0 auto" }}>
                {group.title}
              </Cell>
            </Row>
          )}
          {renderedRow}
          {row.isExpanded && (
            <Row
              {...rowProps}
              className={classNames(rowProps.className, "dp-Table-description")}
              key={`${rowProps.key}_expanded`}
            >
              {renderExpandedRow
                ? renderExpandedRow(row.original as ListItem)
                : null}
            </Row>
          )}
        </React.Fragment>
      );
    }
    return renderedRow;
  };

  return (
    <TableView
      {...getTableProps({
        className: classNames({ "is-hoverable": hoverable })
      })}
    >
      {headerRows.map(headerRow => (
        <HeaderRow {...headerRow.getRowProps()}>
          {headerRow.headers.map((column: EnhancedColumn) => (
            <Header
              {...column.getHeaderProps({
                className: column.headerClassName
              })}
              canSortBy={column.canSortBy}
              sorted={column.sorted}
              sortedDesc={column.sortedDesc}
              sortedIndex={column.sortedIndex}
              sortByProps={column.getSortByToggleProps()}
            >
              {column.id === "checkbox" && (
                <Checkbox
                  id="select_all"
                  checked={selection.length > 0}
                  undef={selection.length > 0 && selection.length < rows.length}
                  onChange={() =>
                    selection.length ? handleSelectNone() : handleSelectAll()
                  }
                />
              )}
              {column.render("Header")}
            </Header>
          ))}
        </HeaderRow>
      ))}
      {!!rows && rows.length > 0 ? rows.map(renderRow) : null}
    </TableView>
  );
};

export default Table;
