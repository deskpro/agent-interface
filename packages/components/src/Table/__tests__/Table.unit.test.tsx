import * as React from "react";
import { render, fireEvent } from "react-testing-library";

import Table from "../Table";

describe("<Table />", () => {
  it("should match its snapshot", () => {
    const { container } = render(
      <Table
        data={[
          { id: 1, title: "First item" },
          { id: 2, title: "Second Title" }
        ]}
        columns={[
          { Header: "ID", accessor: "id", canSortBy: false },
          { Header: "Title", accessor: "title" }
        ]}
        onSort={jest.fn()}
      />
    );
    expect(container.querySelector(".dp-Tableview")).toMatchSnapshot();
  });

  it("should render 4 rows with header", () => {
    const { container } = render(
      <Table
        data={[
          { id: 1, title: "First item" },
          { id: 2, title: "Second Title" },
          { id: 3, title: "Third Title" },
          { id: 4, title: "Fourth Title" }
        ]}
        columns={[
          { Header: "ID", accessor: "id", canSortBy: false },
          { Header: "Title", accessor: "title" }
        ]}
        onSort={jest.fn()}
      />
    );
    expect(container.querySelectorAll(".dp-CustomCheckbox").length).toBe(0);
    expect(container.querySelectorAll(".dp-TableRow").length).toBe(5);
  });

  it("should render checkboxes", () => {
    const { container } = render(
      <Table
        data={[
          { id: 1, title: "First item" },
          { id: 2, title: "Second Title" },
          { id: 3, title: "Third Title" },
          { id: 4, title: "Fourth Title" }
        ]}
        columns={[
          { Header: "ID", accessor: "id", canSortBy: false },
          { Header: "Title", accessor: "title" }
        ]}
        checkable
        onCheckChange={jest.fn()}
        onSort={jest.fn()}
      />
    );
    expect(container.querySelectorAll(".dp-CustomCheckbox").length).toBe(5);
    fireEvent.click(container.querySelector(".dp-CustomCheckbox input"));
    expect(
      container.querySelectorAll(".dp-CustomCheckbox input:checked").length
    ).toBe(5);
    expect(
      container.querySelectorAll(".dp-CustomCheckbox input:not(:checked)")
        .length
    ).toBe(0);
  });

  it("should display sorting state and control", () => {
    const { container } = render(
      <Table
        data={[
          { id: 1, title: "First item" },
          { id: 2, title: "Second Title" },
          { id: 3, title: "Third Title" },
          { id: 4, title: "Fourth Title" }
        ]}
        columns={[
          { Header: "ID", accessor: "id", canSortBy: false },
          { Header: "Title", accessor: "title" }
        ]}
        checkable
        onCheckChange={jest.fn()}
        onSort={jest.fn()}
      />
    );
    expect(container.querySelectorAll(".dp-Icon.dp-arrowGroup").length).toBe(1);

    const sortEl = container.querySelector(".dp-Icon.dp-arrowGroup");
    expect(sortEl.classList.contains("is-asc")).toBeFalsy();
    expect(sortEl.classList.contains("is-desc")).toBeFalsy();

    fireEvent.click(sortEl);
    expect(sortEl.classList.contains("is-asc")).toBeTruthy();
    expect(sortEl.classList.contains("is-desc")).toBeFalsy();
    fireEvent.click(sortEl);
    expect(sortEl.classList.contains("is-asc")).toBeFalsy();
    expect(sortEl.classList.contains("is-desc")).toBeTruthy();
  });
});
