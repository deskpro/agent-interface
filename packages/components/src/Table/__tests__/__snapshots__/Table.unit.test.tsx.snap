// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Table /> should match its snapshot 1`] = `
<div
  class="dp-Tableview is-hoverable"
>
  <div
    class="dp-TableRow dp-TableControl"
    style="display: flex; min-width: 2px;"
  >
    <div
      class="dp-TableItem"
      style="box-sizing: border-box; flex: 1 0 auto; width: 1px; max-width: undefinedpx;"
    >
      ID
    </div>
    <div
      class="dp-TableItem"
      style="box-sizing: border-box; flex: 1 0 auto; width: 1px; max-width: undefinedpx;"
    >
      Title
      <span
        class="dp-Icon dp-arrowGroup"
        style="cursor: pointer;"
        title="Toggle SortBy"
      />
    </div>
  </div>
  <div
    class="dp-TableRow"
    style="display: flex; min-width: 2px;"
  >
    <div
      class=""
      style="display: block; box-sizing: border-box; flex: 1 0 auto; width: 1px; max-width: undefinedpx;"
    >
      1
    </div>
    <div
      class=""
      style="display: block; box-sizing: border-box; flex: 1 0 auto; width: 1px; max-width: undefinedpx;"
    >
      First item
    </div>
  </div>
  <div
    class="dp-TableRow"
    style="display: flex; min-width: 2px;"
  >
    <div
      class=""
      style="display: block; box-sizing: border-box; flex: 1 0 auto; width: 1px; max-width: undefinedpx;"
    >
      2
    </div>
    <div
      class=""
      style="display: block; box-sizing: border-box; flex: 1 0 auto; width: 1px; max-width: undefinedpx;"
    >
      Second Title 
    </div>
  </div>
</div>
`;
